File: Templates y codigos curiosos/Recien traducido/Matematica/SieveSundaram.cpp

#include <bits/stdc++.h> 
using namespace std;

int SieveOfSundaram(int n) {

    int nNew = (n - 1) / 2;
    bool marked[nNew + 1];
    memset(marked, false, sizeof (marked));
    for (int i = 1; i <= nNew; i++)
        for (int j = i; (i + j + 2 * i * j) <= nNew; j++)
            marked[i + j + 2 * i * j] = true;
    if (n > 2)
        cout << 2 << " ";
    for (int i = 1; i <= nNew; i++)
        if (marked[i] == false)
            cout << 2 * i + 1 << " ";
}

int main(void) {
    int n = 20;
    SieveOfSundaram(n);
    return 0;
} 

File: Codigos_por_competidor/Diego_Rodriguez/Algoritmos/SieveSundaram.kt
import kotlin.jvm.JvmStatic
import java.util.Arrays

object SieveSundaram {
    // Prints all prime numbers smaller 
    fun SieveOfSundaram(n: Int): Int {

        // In general Sieve of Sundaram, produces  
        // primes smaller than (2*x + 2) for a number 
        // given number x. Since we want primes  
        // smaller than n, we reduce n to half 
        val nNew = (n - 1) / 2

        // This array is used to separate numbers of the  
        // form i+j+2ij from others where 1 <= i <= j 
        val marked = BooleanArray(nNew + 1)

        // Initialize all elements as not marked 
        Arrays.fill(marked, false)

        // Main logic of Sundaram. Mark all numbers of the 
        // form i + j + 2ij as true where 1 <= i <= j 
        for (i in 1..nNew) {
            var j = i
            while (i + j + 2 * i * j <= nNew) {
                marked[i + j + 2 * i * j] = true
                j++
            }
        }

        // Since 2 is a prime number 
        if (n > 2) {
            print(2.toString() + " ")
        }

        // Print other primes. Remaining primes are of  
        // the form 2*i + 1 such that marked[i] is false. 
        for (i in 1..nNew) {
            if (marked[i] == false) {
                print((2 * i + 1).toString() + " ")
            }
        }
        return -1
    }

    // Driver code 
    @JvmStatic
    fun main(args: Array<String>) {
        val n = 20
        SieveOfSundaram(n)
    }
}